### PR TITLE
improvements & implementations for Mesh

### PR DESCRIPTION
**Pull request by IlliaHryshchuk**
Completed:
 - Read/write methods for Mesh

Added:
 - getNormal for Triangle
 - unit tests for Mesh and Triangle

### CODE REVIEW

**Comment by IlliaHryshchuk** [comment]
why does it happens? Should i fetch & merge from development, and implement my changes one more time?

image

File: GeometryCore/include/Mesh.h [hunk 1]
```diff
@@ -8,17 +9,29 @@ namespace geom_utils
 {
     class Mesh
     {
-        std::vector<FPoint3D> vertices;
+        std::vector<Triangle3D> vertices;
+        std::string comment;
```
Comment (VladGribinchuk) [inline GeometryCore/include/Mesh.h]: What intention/goal comment has?

File: GeometryCore/include/Triangle.h [hunk 2]
```diff
@@ -15,10 +14,14 @@ namespace geom_utils
...
     {
+        Triangle3D() {};
         Triangle3D(const FPoint3D& a, const FPoint3D& b, const FPoint3D& c)
             : a(a), b(b), c(c)
         {}
 
         FPoint3D a, b, c;
     };
+
+    FPoint3D getNormal(const Triangle3D& triangel);
```
Comment (VladGribinchuk) [inline GeometryCore/include/Triangle.h]: Make this function as Triangle3D method instead.
FPoint3D Triangle3D::normal() const {...}

File: GeometryCore/include/Triangle.h [hunk 3]
```diff
@@ -15,10 +14,14 @@ namespace geom_utils
...
+    FPoint2D getNormal(const Triangle2D& triangel);
```
Comment (VladGribinchuk) [inline GeometryCore/include/Triangle.h]: I'm thinking... that normal for 2D triangle in 2D has no sense.. So remove this

File: GeometryCore/src/Triangle.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#include "../include/LinearAlgebra.h"
+#include "../include/Triangle.h"
+
+namespace geom_utils
+{
+
+	FPoint3D getNormal(const Triangle3D& triangel)
+	{
+		//compute vector
+		FPoint3D normal = geom_utils::cross(triangel.a, triangel.b);
```
Comment (VladGribinchuk) [inline GeometryCore/src/Triangle.cpp]: Don't think this is absolutely correct.
Firstly, you should calsulate vectors: vecA = b - a; vecB = c - a; Then vecA x vecB.

File: GeometryCore/src/Triangle.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		float length = sqrt(normal.x * normal.x + normal.y * normal.y + normal.z * normal.z);
+		normal.x = normal.x / length;
+		normal.y = normal.y / length;
+		normal.z = normal.z / length;
+
+		return normal;
+	};
+
+
+	FPoint2D getNormal(const Triangle2D& triangel)
```
Comment (VladGribinchuk) [inline GeometryCore/src/Triangle.cpp]: As I mentioned this doesn't make sense, so remove

File: UnitTests/MeshTests.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        ), "The read mesh is wrong!");
+}
+
+
+DEFINE_TEST_CASE(MeshWrite)
+{
+    Mesh mesh;
+    mesh.add({ {1, 1, 0 }, {0, 1, 0}, {0, 1, -1} });
+    mesh.setComment("square_unit_test");
+	MeshHandler::write("../test_models/square_unit_WriteTest_ascii.stl", mesh);
```
Comment (VladGribinchuk) [inline UnitTests/MeshTests.h]: Ok, but remove any temporary artifacts at the end of test case.
`std::remove("../test_models/square_unit_WriteTest_ascii.stl");`

File: UnitTests/UnitTests.vcxproj [hunk 7]
```diff
@@ -88,6 +88,7 @@
       <PreprocessorDefinitions>_DEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <ConformanceMode>true</ConformanceMode>
       <OpenMPSupport>true</OpenMPSupport>
+      <AdditionalIncludeDirectories>C:\Users\asus\source\repos\VladGribinchuk\FoundationBuilder\GeometryCore;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Comment (VladGribinchuk) [inline UnitTests/UnitTests.vcxproj]: Here is the absolute path. If you really need this additional include directory, then use only relative path

File: UnitTests/UnitTests.vcxproj.filters [hunk 8]
```diff
@@ -1,36 +1,16 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
-    <Filter Include="Source Files">
```
Comment (VladGribinchuk) [inline UnitTests/UnitTests.vcxproj.filters:4]: Probably, you've accidently include these scope of changes. 
Just make another commit with restoring previous state of this file.

File: test_models/square_unit_WriteTest_ascii.stl [hunk 9]
```diff
@@ -0,0 +1,10 @@
+solid square_unit_test
```
Comment (VladGribinchuk) [inline test_models/square_unit_WriteTest_ascii.stl]: Don't commit this test artifact file

**Review by VladGribinchuk — CHANGES_REQUESTED** [review]

**Comment by VladGribinchuk** [comment]
> why does it happens? Should i fetch & merge from development, and implement my changes one more time?
> 
> image

See one of my comments to the PR

File: UnitTests/UnitTests.vcxproj.filters (hunk 8 above)
Comment (VladGribinchuk) [reply UnitTests/UnitTests.vcxproj.filters:4]: Clarification: maybe at some point project filters were broken in UnitTests project. You can make those filters again. Search project in Solution Explorer, Add->New Filter. Name one filter - Header Files and another filter - Source Files. Drag files in this project under corresponding filter. Save and commit.
This is not a huge problem, but fix would help to keep project tidy.

File: GeometryCore/include/Mesh.h [hunk 10]
```diff
@@ -8,17 +9,29 @@ namespace geom_utils
 {
     class Mesh
     {
-        std::vector<FPoint3D> vertices;
+        std::vector<Triangle3D> vertices;
```
Comment (VladGribinchuk) [inline GeometryCore/include/Mesh.h]: This is not vertices now. More appropriate would be triangles or facets.

File: GeometryCore/include/Mesh.h (hunk 1 above)
Comment (IlliaHryshchuk) [reply GeometryCore/include/Mesh.h]: it contains comment of stl file, I thought it would be useful. Should i remove it?

image

File: GeometryCore/include/Triangle.h (hunk 2 above)
Comment (IlliaHryshchuk) [reply GeometryCore/include/Triangle.h]: done

File: GeometryCore/include/Triangle.h (hunk 3 above)
Comment (IlliaHryshchuk) [reply GeometryCore/include/Triangle.h]: done

File: GeometryCore/src/Triangle.cpp (hunk 4 above)
Comment (IlliaHryshchuk) [reply GeometryCore/src/Triangle.cpp]: done

File: UnitTests/MeshTests.h (hunk 6 above)
Comment (IlliaHryshchuk) [reply UnitTests/MeshTests.h]: Ok, i did not know about it. It is done

File: UnitTests/UnitTests.vcxproj (hunk 7 above)
Comment (IlliaHryshchuk) [reply UnitTests/UnitTests.vcxproj]: It was created accidentally, i will try to remove it

File: UnitTests/UnitTests.vcxproj.filters (hunk 8 above)
Comment (IlliaHryshchuk) [reply UnitTests/UnitTests.vcxproj.filters:4]: Done. As i understood, that filters are something like simple folders?

File: test_models/square_unit_WriteTest_ascii.stl (hunk 9 above)
Comment (IlliaHryshchuk) [reply test_models/square_unit_WriteTest_ascii.stl]: Ok, i will remove it

File: GeometryCore/include/Mesh.h (hunk 10 above)
Comment (IlliaHryshchuk) [reply GeometryCore/include/Mesh.h]: fixed

File: GeometryCore/include/Mesh.h (hunk 1 above)
Comment (VladGribinchuk) [reply GeometryCore/include/Mesh.h]: I see... then maybe rename this to `std::string modelName`; smth like this

File: UnitTests/UnitTests.vcxproj.filters (hunk 8 above)
Comment (VladGribinchuk) [reply UnitTests/UnitTests.vcxproj.filters:4]: Yeap

File: GeometryCore/include/Mesh.h (hunk 10 above)
Comment (VladGribinchuk) [reply GeometryCore/include/Mesh.h]: Have you pushed your fixes? coz i don't see

File: GeometryCore/include/Mesh.h (hunk 1 above)
Comment (IlliaHryshchuk) [reply GeometryCore/include/Mesh.h]: ok, i will rename it

File: GeometryCore/include/Mesh.h (hunk 10 above)
Comment (IlliaHryshchuk) [reply GeometryCore/include/Mesh.h]: it is not pushed yet, because i have not fixed the trouble with absolute path. All changes are local 

upd: pushed

File: UnitTests/UnitTests.vcxproj (hunk 7 above)
Comment (IlliaHryshchuk) [reply UnitTests/UnitTests.vcxproj]: upd:
seems like it is fixed

File: FoundationBuilder/FoundationBuilder.vcxproj [hunk 11]
```diff
@@ -100,6 +100,7 @@
       <PreprocessorDefinitions>_DEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <ConformanceMode>true</ConformanceMode>
       <OpenMPSupport>true</OpenMPSupport>
+      <AdditionalIncludeDirectories>C:\Users\asus\source\repos\VladGribinchuk\FoundationBuilder\GeometryCore;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Comment (VladGribinchuk) [inline FoundationBuilder/FoundationBuilder.vcxproj]: This is still absolute path in FoundationBuilder project

Comment (VladGribinchuk) [reply FoundationBuilder/FoundationBuilder.vcxproj]: If this is not necessary change - you can simply open this file by notepad and just remove this line)

**Comment by VladGribinchuk** [comment]
When you deal with the last fix, there will be one more sub-task in the scope of Mesh functionality.
Please check out my https://github.com/VladGribinchuk/FoundationBuilder/tree/VG-unit-tests branch. This branch contains my version of Mesh/MeshHandler implementation.
Your code is good and ready to be merged into our main development branch. But please review how this functionality is impelemented in VG-unit-tests. There you can find the callbacks system example - architecture approach used in some C++ projects (pay close attention to lambda expressions usage, std::function objects and special std::bind object). It can be useful for tracking and logging how reading/writing operation are going, and output to console any warning or errors occured during these operations.
So, the goal is:
- checkout VG-unit-tests, review and understand how reading/writing mesh is working
- compare robustness and speed of read/write mesh operations between dev1_Illia-Mesh_class-#19 and VG-unit-tests branches.
- unite(manually copy and paste) functionality from both branches (for example, find additional checks during reading/writing operations implemented in VG-unit-tests your; also your branch provides mesh writing in ascii format, but VG-unit-tests provides writing in binary format, and so on).

Code in VG-unit-tests was implemented with learning purpose. But you can try to implement similar callbacks system in your branch. It may be handy in future in FoundationBuilder project for logging information about application progress during its running. I'm not insist, just suggest.

**Comment by IlliaHryshchuk** [comment]
I did not implement the binary writing becouse I don't know how to use it in the project (we don't have a flag for binar/ascii writing), but i can add it if it is necessary, or it would be usefull in the future.
Ok, i will check VG-unit-tests, it is interesting for me) And i will commit changes after that (may be i will add something for my current pull request)

**Comment by VladGribinchuk** [comment]
Please dont consider all my preparations as a hard rule. You are free to modify classes/methods if you see possibility for extending or improving. I have implemented base classes just to give the general idea for the team how to move on.
So, surely, you can add additional boolean argument as a flag for switching ascii/binary format for writing. And yes, if you will make additional changes, please commit them into your current branch. Anyway let me please know when you finish.

**Comment by IlliaHryshchuk** [comment]
What does that `using` mean? it means that the memory after execution of this part of code could be free?

image

**Comment by VladGribinchuk** [comment]
> What does that `using` mean? it means that the memory after execution of this part of code could be free?
> 
> image

No, things are more simpler) `using` in this contex is just type alias, i.e. identiacal to c++ `typedef`.
It means that instead of the following declaration of functionObject
`std::function<void(std::string)> functionObject;`
we can just type:
`TraceCallback_t functionObject;`

Compare these two identical variants above. Second variant is more easier and pretty, while logical loading is the same and programmist can type less number of symbols:)

**Comment by IlliaHryshchuk** [comment]
> > What does that `using` mean? it means that the memory after execution of this part of code could be free?
> > image
> 
> No, things are more simpler) `using` in this contex is just type alias, i.e. identiacal to c++ `typedef`.
> It means that instead of the following declaration of functionObject
> `std::function<void(std::string)> functionObject;`
> we can just type:
> `TraceCallback_t functionObject;`
> 
> Compare these two identical variants above. Second variant is more easier and pretty, while logical loading is the same and programmist can type less number of symbols:)

Yes, i have understood, thanx)

**Comment by IlliaHryshchuk** [comment]
I have added some features from `VG-unit-tests`, but i have not added the callbacks system yet, because i have a lot of questions how does it work)

**Comment by VladGribinchuk** [comment]
Changes look good! What questions do you have, btw?

**Comment by IlliaHryshchuk** [comment]
While i was writting all my questions, i have accidentally understood how does it work :) So i can add callbacks now.
But i still don't understand something:

 - why the `mesh` object is transfered to lambda with `Traingle3D` object? And why we have to use `std::bind` here?
 - `Mesh::add` has only 1 parameter, so why do we use `std::placeholdres::_1` and `&mesh` ? it means that every `Mesh::add` will use only first parameter from brackets, or it will use only `&mesh` ? 

image

 - `trace` is initialized by `[this](std::string msg) { TEST_ASSERT(false, "Smth was wrong: " + msg); /*unexpected issue*/ },` - so it means that every time when the `trace` is called - unit tests have to fail?
 - and is it unnecessary to check every field of written/read mesh in the unit tests?

**Comment by VladGribinchuk** [comment]
> While i was writting all my questions, i have accidentally understood how does it work

Haha, here is the true my friend: answer lies in the good formulated question.

I assume you have familiarized with [std::bind](http://www.cplusplus.com/reference/functional/bind/). Please see my answers below:
- bind is used here in order to create callable object (std::function which wraps Mesh::add function). Mesh::add is a class member function. In C++ class member function is the same as free (non-member) function, except that **compiler implicitly passes** the pointer to object as its first argument (you are able to access it inside method as `this`), and this is the key. There must be an object to call its method, right? 
So, in this example, we pass Mesh::add method as callback. And we tie its first argument to pointer to Mesh object `&mesh`. We have to do this because Mesh::add is a class member function. In this case, compiler won't do this implicitly for us, so we must manually specify this. Then we use `placeholders::_1` to specify that first argument of the function call should be passed as second argument to Mesh::add (remember, first argument in `bind` is &mesh, because we bind class member function). 
You can see that first argument of the function call is Triangle3D and it is being passed to `mesh` object in the same way as we just call `mesh.add(triangle);`.
Compare: if add() was non-member function, then we would just `std::bind(&add, std::placeholder::_1);`
- the first answer covers your second question. In real `Mesh::add` has 2 parameters. First is implicitly passed by compiler - the pointer to the object which calls method add(). Second parameter is being explicitly specified by developer. 
But since `bind` is not standard call, e.g. `mesh.add(triangle)`, compiler doesn't know which object should call method add(), so we have to specify object manually.
- Yes, in my implementation `trace` is called when something in read/write process goes wrong. So `TEST_ASSERT(false`) forces test case to fail.
- I cannot answer definitely... but here is the principle: the more detailed test case is, the more control over the code you have; the larger code coverage by unit tests, the less unexpected issues you have.

**Comment by VladGribinchuk** [comment]
image

And don't forget to resolve conflicts of this branch, because it cannot be merged if there are conflicts. 
Here is the algorithm: pull the latest changes of development branch, merge development into your current working branch, resolve merge conflicts, commit, push.

**Comment by IlliaHryshchuk** [comment]
> I assume you have familiarized with [std::bind](http://www.cplusplus.com/reference/functional/bind/).

yes, but it was more clear with simple examples from youtube and cplusplus.com. There were similar number of parameters, and simple functions )

> * In real `Mesh::add` has 2 parameters. First is implicitly passed by compiler - the pointer to the object which calls method add().

Oh, i didn't know it. So now it is clear why there are 2 parameters.

**Comment by IlliaHryshchuk** [comment]
seems that conflicts are resolved

**Comment by VladGribinchuk** [comment]
Yes, now conflicts are resolved and Im ready to approve and merge this PR. Do you plan to commit something else or we can merge?

**Comment by IlliaHryshchuk** [comment]
There is not much time left, and i'm afraid that we will have to hurry up, so you can merge it. But i can create new pull request and add callbacks later, after all issues will be done, right?

**Comment by VladGribinchuk** [comment]
Surely

**Review by VladGribinchuk — APPROVED** [review]
